src/interpreter/symbolTable.ts

```diff
@@ -1,88 +1,248 @@
+        throw new UnsupportedDeclarationError();
+      }
     }
-    symbolTable.head[mapping.name] = {
-      address: mapping.address,
-      nameType: mapping.nameType
-    };
   });
+  return symbolTable;
 };
 
-/**
- * Adds a new frame to the symbol table. Each frame represents a different scope.
- */
-export const extendSymbolTableWithEntries = (
-  declarationNames: DeclarationNameWithAddress[],
+export const addFunctionSymbolTableEntries = (
+  paramDeclarations: VariableDeclaration[],
+  node: FunctionDeclaration,
   symbolTable: SymbolTable
 ): SymbolTable => {
-  const newFrame: SymbolTableFrame = {};
+  const frame: SymbolTableFrame = {};
+  // Params offset from below the rbp, start from -2 to leave a space for return address.
+  let offset = -2;
+  paramDeclarations.forEach((declaration) => {
+    const entries = constructVariableDeclarationSymbolTableEntries(
+      declaration,
+      'Function',
+      offset
+    );
+    addToFrame(frame, ...entries);
+    offset -= entries.length;
+  });
 
-  declarationNames.forEach((mapping) => {
-    if (mapping.name in newFrame) {
-      throw new RedeclaredNameError(
-        'Tried to redeclare a name in the same scope.'
-      );
-    }
-    newFrame[mapping.name] = {
-      address: mapping.address,
-      nameType: mapping.nameType
-    };
+  offset = 0;
+  const bodyDeclarations = !isEmptyStatement(node.body)
+    ? node.body.items.filter((item): item is VariableDeclaration =>
+        isVariableDeclaration(item)
+      )
+    : [];
+  bodyDeclarations.forEach((declaration) => {
+    const entries = constructVariableDeclarationSymbolTableEntries(
+      declaration,
+      'Function',
+      offset
+    );
+    addToFrame(frame, ...entries);
+    offset += entries.length;
   });
 
+  const functionEntry = getFunctionSymbolTableEntry(node.id.name, symbolTable);
+  functionEntry.numOfVariables += offset;
+
   return {
-    head: newFrame,
-    tail: symbolTable
+    head: frame,
+    tail: symbolTable,
+    parent: functionEntry
   };
 };
 
-/**
- * Gets the entry of a name.
- */
-export const getEntryFromSymbolTable = (
+export const addBlockSymbolTableEntries = (
+  block: BlockStatement,
+  symbolTable: SymbolTable
+): SymbolTable => {
+  const frame: SymbolTableFrame = {};
+  if (!isNotNull(symbolTable.parent)) {
+    throw new InvalidScopeError('Block is not inside a function.');
+  }
+  let offset = symbolTable.parent.numOfVariables;
+  const declarations = block.items.filter((item): item is VariableDeclaration =>
+    isVariableDeclaration(item)
+  );
+  declarations.forEach((declaration) => {
+    const entries = constructVariableDeclarationSymbolTableEntries(
+      declaration,
+      'Block',
+      offset
+    );
+    addToFrame(frame, ...entries);
+    offset += entries.length;
+  });
+
+  symbolTable.parent.numOfVariables +=
+    offset - symbolTable.parent.numOfVariables;
+
+  return {
+    head: frame,
+    tail: symbolTable,
+    parent: symbolTable.parent
```

Comment:
Am I right to say that the whole point of `parent` is so that blocks can access the offset in which they should start with, regardless of how nested they are & also to deal with sequential blocks:
```c
{
    ...
}
...
{
    ...
}
```

Reply:
Yes